lib/services/local/mfa_test.go

```diff
@@ -0,0 +1,231 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package local_test
+
+import (
+	"cmp"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	mfav1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/mfa/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend/memory"
+	local "github.com/gravitational/teleport/lib/services/local"
+)
+
+func TestMFAService_CRUD(t *testing.T) {
+	t.Parallel()
+
+	backend, err := memory.New(memory.Config{})
+	require.NoError(t, err)
+
+	svc, err := local.NewMFAService(backend)
+	require.NoError(t, err)
+
+	username, chal := "alice", newValidatedMFAChallenge()
+
+	// Create
+	created, err := svc.CreateValidatedMFAChallenge(t.Context(), username, chal)
+	require.NoError(t, err)
+	require.Equal(t, chal.Kind, created.Kind)
+	require.Equal(t, chal.Version, created.Version)
+	require.Equal(t, chal.Metadata.Name, created.Metadata.Name)
+	require.WithinDuration(t, time.Now().Add(5*time.Minute), *created.Metadata.Expires, time.Second)
+
+	// Get
+	got, err := svc.GetValidatedMFAChallenge(t.Context(), username, chal.Metadata.Name)
+	require.NoError(t, err)
+	require.Equal(t, created.Metadata.Name, got.Metadata.Name)
+	require.Equal(t, created.Spec.SourceCluster, got.Spec.SourceCluster)
+	require.Equal(t, created.Spec.TargetCluster, got.Spec.TargetCluster)
+	require.WithinDuration(t, *created.Metadata.Expires, *got.Metadata.Expires, time.Second)
+
+	// Get non-existent
+	_, err = svc.GetValidatedMFAChallenge(t.Context(), username, "does-not-exist")
+	require.ErrorContains(t, err, "doesn't exist")
+}
+
+func TestMFAService_CreateValidatedMFAChallenge_Validation(t *testing.T) {
+	t.Parallel()
+
+	backend, err := memory.New(memory.Config{})
+	require.NoError(t, err)
+
+	svc, err := local.NewMFAService(backend)
+	require.NoError(t, err)
+
+	defaultUsername := "bob"
+
+	for _, testCase := range []struct {
+		name     string
+		username *string
+		chal     *mfav1.ValidatedMFAChallenge
+		wantErr  bool
+	}{
+		{
+			name:    "valid challenge",
+			chal:    newValidatedMFAChallenge(),
+			wantErr: false,
+		},
+		{
+			name: "missing username",
+			username: func() *string {
+				emptyStr := ""
+				return &emptyStr
+			}(),
+			wantErr: true,
+		},
+		{
+			name:    "nil challenge",
+			chal:    nil,
+			wantErr: true,
+		},
+		{
+			name: "invalid kind",
+			chal: &mfav1.ValidatedMFAChallenge{
+				Kind:     "wrong_kind",
+				Version:  "v1",
+				Metadata: &types.Metadata{Name: "test"},
+				Spec: &mfav1.ValidatedMFAChallengeSpec{
+					Payload: &mfav1.SessionIdentifyingPayload{
+						Payload: &mfav1.SessionIdentifyingPayload_SshSessionId{SshSessionId: []byte("id")},
+					},
+					SourceCluster: "src", TargetCluster: "tgt",
+				},
+			},
```

Comment:
Modify a valid challenge, so we only see in the test what differs.

```suggestion
			chal: func() *mfav1.ValidatedMFAChallenge{
				c := newValidatedMFAChallenge()
				c.Kind = "wrong_kind"
				return c
			}(),
```

There are many ways to do this, the anonymous func is just one of them.

Same for the other tests.